### PR TITLE
fix(dropdown): browser collision detection

### PR DESCRIPTION
**Pull request by SjorsH**
Fix for the dropdown issues #5841, #5767

### CODE REVIEW

File: src/dropdown/dropdown.js [hunk 1]
```diff
@@ -187,17 +187,31 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+    if (clientRect.top > dropdownHeight &&
+        window.innerHeight - clientRect.top < clientRect.height + dropdownHeight) {
+      return true;
+    }
+    return false;
+  }
+
   scope.$watch('isOpen', function(isOpen, wasOpen) {
     if (appendTo && self.dropdownMenu) {
       var pos = $position.positionElements($element, self.dropdownMenu, 'bottom-left', true),
```
Comment (RobJacobs) [inline src/dropdown/dropdown.js]: The placement parameter passed to the positionElements method should be changed to 'auto bottom-left' and leverage the work already being done there.

File: src/dropdown/dropdown.js [hunk 2]
```diff
@@ -187,17 +187,30 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+      var pos = $position.positionElements($element, self.dropdownMenu, 'auto bottom-left', true),
         rightalign,
-        scrollbarWidth;
-
-      css = {
-        top: pos.top + 'px',
-        display: isOpen ? 'block' : 'none'
-      };
+        scrollbarWidth,
+        css = {
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: Grouping the css setting of this object hurts readability - better to move it back to how it was prior.

File: src/dropdown/dropdown.js [hunk 3]
```diff
@@ -187,17 +187,30 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
     }
   };
 
+  function toggleDropdownUp() {
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: This should be abstracted into a function into the $position service

Comment (SjorsH) [inline src/dropdown/dropdown.js]: Thanks, I'm going back to the drawing board and will let you know if I changed the position service.

Comment (RobJacobs) [inline src/dropdown/dropdown.js]: There should be no need to change the position service to accomplish this.  Calling the $position.positionElements with an auto placement should do what you need, just use the pos.top value returned.

Comment (SjorsH) [inline src/dropdown/dropdown.js]: @RobJacobs 
Indeed it works fine with placement set to auto, but I noticed that width and height of my targetElement inside the $position.positionElements are 0.

`// need to read from prop to support tests.
        var targetWidth = angular.isDefined(targetElem.offsetWidth) ? targetElem.offsetWidth : targetElem.prop('offsetWidth');
        var targetHeight = angular.isDefined(targetElem.offsetHeight) ? targetElem.offsetHeight : targetElem.prop('offsetHeight');`

Why not use the outerWidth and outerHeight?

Comment (RobJacobs) [inline src/dropdown/dropdown.js]: Because outerHieght and outerWidth are jQuery methods and we only support jQLite or native element methods and properties.  Besides, the outerHeight/width jQuery documentation states:

Get the current computed height for the first element in the set of matched elements, including padding, border, and optionally margin.

And the HtmlElement offsetHeight/width documentation from MDN states:

Typically, an element's offsetHeight is a measurement which includes the element borders, the element vertical padding, the element horizontal scrollbar (if present, if rendered) and the element CSS height.

So the 2 are the same unless you include the margin parameter.

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "author": "https://github.com/angular-ui/bootstrap/graphs/contributors",
   "name": "angular-ui-bootstrap",
-  "version": "1.3.3-SNAPSHOT",
+  "version": "1.3.3",
```
Comment (wesleycho) [inline package.json]: This change needs to be undone

File: src/dropdown/dropdown.js [hunk 5]
```diff
@@ -189,7 +189,7 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
 
   scope.$watch('isOpen', function(isOpen, wasOpen) {
     if (appendTo && self.dropdownMenu) {
-      var pos = $position.positionElements($element, self.dropdownMenu, 'bottom-left', true),
+      var pos = $position.positionElements($element, self.dropdownMenu, 'auto bottom-left', true),
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: What is the rationale for this change? This changes the default for all apps using this component.

File: src/position/position.js [hunk 6]
```diff
@@ -436,8 +436,8 @@ angular.module('ui.bootstrap.position', [])
         targetElem = this.getRawNode(targetElem);
 
         // need to read from prop to support tests.
-        var targetWidth = angular.isDefined(targetElem.offsetWidth) ? targetElem.offsetWidth : targetElem.prop('offsetWidth');
-        var targetHeight = angular.isDefined(targetElem.offsetHeight) ? targetElem.offsetHeight : targetElem.prop('offsetHeight');
+        var targetWidth = angular.element(targetElem).outerWidth();
+        var targetHeight = angular.element(targetElem).outerHeight();
```
Comment (wesleycho) [inline src/position/position.js]: @RobJacobs how do you feel about this change here?

Comment (wesleycho) [inline src/position/position.js]: Actually, it looks like this change is breaking the tests

Comment (RobJacobs) [inline src/position/position.js]: Those are jQuery methods (outerWidth and outerHeight) and not native element or jQlite methods.  The note above the line of code removed says the prop lookups are there for a reason - to support the tests, so I'm not sure why they are removed.  The position test creates a mock element [here](https://github.com/angular-ui/bootstrap/blob/master/src/position/test/position.spec.js#L6) with those values as props.

Comment (wesleycho) [inline src/position/position.js]: Thanks, looks like this change needs to be undone.

**Comment by wesleycho** [comment]
Going to close this due to fundamental issues - I'll take a stab at this soon at a proper implementation.
